src/java/org/apache/cassandra/index/sai/disk/v2/sortedterms/SortedTermsReader.java

```diff
@@ -0,0 +1,303 @@
+        RandomAccessInput offsetSlice = termsDataBlockOffsets.randomAccessSlice(0, termsDataBlockOffsets.length());
+        this.blockOffsets = DirectMonotonicReader.getInstance(offsetsMeta, offsetSlice);
+    }
+
+    /**
+     * Returns the point id (ordinal) of the target term or the next greater if no exact match found.
+     * If reached the end of the terms file, returns <code>Long.MAX_VALUE</code>.
+     * Complexity of this operation is O(log n).
+     *
+     * @param term target term to lookup
+     * @throws IOException if a read of the internal on-disk trie fails
+     */
+    public long getPointId(@Nonnull ByteComparable term) throws IOException
+    {
+        Preconditions.checkNotNull(term, "term null");
+
+        try (TrieRangeIterator reader = new TrieRangeIterator(termsTrie.instantiateRebufferer(),
+                                                              meta.trieFP,
+                                                              term,
+                                                              null,
+                                                              true,
+                                                              true))
+        {
+            final Iterator<Pair<ByteSource, Long>> iterator = reader.iterator();
+            return iterator.hasNext() ? iterator.next().right : Long.MAX_VALUE;
+        }
+    }
+
+    /**
+     * Returns the total number of terms.
+     */
+    public long count()
+    {
+        return meta.count;
+    }
+
+    /**
+     * Opens a cursor over the terms stored in the terms file.
+     * <p>
+     * This does not read any data yet.
+     * The cursor is initially positioned before the first item.
+     * <p>
+     * The cursor is to be used in a single thread.
+     * The cursor is valid as long this object hasn't been closed.
+     * You must close the cursor when you no longer need it.
+     */
+    public @Nonnull Cursor openCursor() throws IOException
+    {
+        return new Cursor(termsData);
+    }
+
+    /**
+     * Allows reading the terms from the terms file.
+     * Can quickly seek to a random term by <code>pointId</code>.
+     * <p>
+     * This object is stateful and not thread safe.
+     * It maintains a position to the current term as well as a buffer that can hold one term.
+     */
+    @NotThreadSafe
+    public class Cursor implements AutoCloseable
+    {
+        private final IndexInput termsData;
+
+        // The term the cursor currently points to. Initially empty.
+        private final BytesRef currentTerm;
+
+        // The point id the cursor currently points to. -1 means before the first item.
+        private long pointId = -1;
+
+        Cursor(IndexInput termsData) throws IOException
+        {
+            this.termsData = termsData.clone();  // clone needed to get our own file pointer independently of other Cursors
```

Comment:
I have a minor concern about this. One of the advantages of the `Ref` / `RefCounted.Tidy` combination is that any resource leaks are reported. I'm not that hung up about this as long we are careful with its usage. I was trying to think if there was a simple way that we could rework `SharedInput` to use the clone operation. I know the original implementation of that class was flawed but if we used `clone` in it, it would overcome those flaws.

Reply:
I'll try to.